Source/WebCore/fileapi/Blob.cpp

```diff
@@ -55,21 +57,56 @@ WTF_MAKE_ISO_ALLOCATED_IMPL(Blob);
 
 class BlobURLRegistry final : public URLRegistry {
 public:
-    void registerURL(ScriptExecutionContext&, const URL&, URLRegistrable&) final;
-    void unregisterURL(const URL&) final;
+    void registerURL(const ScriptExecutionContext&, const URL&, URLRegistrable&) final;
+    bool unregisterURL(const URL&) final;
+    void unregisterURLsForContext(const ScriptExecutionContext&) final;
 
     static URLRegistry& registry();
+
+    Lock m_urlsPerContextLock;
+    HashMap<ScriptExecutionContextIdentifier, HashSet<URL>> m_urlsPerContext WTF_GUARDED_BY_LOCK(m_urlsPerContextLock);
```

Comment:
We might want to make a type for maps of sets this includes the removal algorithm that handles the set becoming empty. Might even optimize that case of removing the last item in the set so we don’t first remove from the hash table and then destroy it.

Reply:
I think this would be nice (though likely separately from this patch). I have definitely written such logic quite a few times so having something reusable would be nice. However, it may need to be templated because the nested type is not always a HashSet<>. I remember writing code where the inner type was a Vector for example.